lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -49,44 +49,74 @@
 namespace Org.Apache.REEF.IMRU.OnREEF.Driver
 {
     /// <summary>
-    /// Implements the IMRU driver on REEF
+    /// Implements the IMRU driver on REEF with fault tolerant
     /// </summary>
     /// <typeparam name="TMapInput">Map Input</typeparam>
     /// <typeparam name="TMapOutput">Map output</typeparam>
     /// <typeparam name="TResult">Result</typeparam>
     /// <typeparam name="TPartitionType">Type of data partition (Generic type in IInputPartition)</typeparam>
-    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType> 
-        : IObserver<IDriverStarted>,
+    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType> :
+        IObserver<IDriverStarted>,
         IObserver<IAllocatedEvaluator>,
         IObserver<IActiveContext>,
         IObserver<ICompletedTask>,
         IObserver<IFailedEvaluator>,
         IObserver<IFailedContext>,
-        IObserver<IFailedTask>
+        IObserver<IFailedTask>,
+        IObserver<IRunningTask>,
+        IObserver<int>
     {
         private static readonly Logger Logger =
             Logger.GetLogger(typeof(IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>));
 
         private readonly ConfigurationManager _configurationManager;
         private readonly int _totalMappers;
-        private readonly IEvaluatorRequestor _evaluatorRequestor;
-        private ICommunicationGroupDriver _commGroup;
         private readonly IGroupCommDriver _groupCommDriver;
-        private readonly TaskStarter _groupCommTaskStarter;
-        private readonly ConcurrentStack<IConfiguration> _perMapperConfiguration;
-        private readonly int _coresPerMapper;
-        private readonly int _coresForUpdateTask;
-        private readonly int _memoryPerMapper;
-        private readonly int _memoryForUpdateTask;
+        private ConcurrentStack<IConfiguration> _perMapperConfigurationStack;
         private readonly ISet<IPerMapperConfigGenerator> _perMapperConfigs;
-        private readonly ISet<ICompletedTask> _completedTasks = new HashSet<ICompletedTask>();
-        private readonly int _allowedFailedEvaluators;
-        private int _currentFailedEvaluators = 0;
         private readonly bool _invokeGC;
-        private int _numberOfReadyTasks = 0;
+        private readonly ServiceAndContextConfigurationProvider<TMapInput, TMapOutput, TPartitionType> _serviceAndContextConfigurationProvider;
 
-        private readonly ServiceAndContextConfigurationProvider<TMapInput, TMapOutput, TPartitionType>
-            _serviceAndContextConfigurationProvider;
+        /// <summary>
+        /// The lock for the driver. 
+        /// </summary>
+        private readonly object _lock = new object();
```

Comment:
ActiveContextManager, EvaluatorManager and TaskManager all marked as   [NotThreadSafe]. This was the decision we made during the code review for those PRs. We would like to have a big lock in driver to avoid nested locks, cross use, and potential dead locks.